lxd/storage/drivers/driver_pure_util.go

```diff
@@ -517,6 +572,475 @@ func (p *pureClient) disconnectHostFromVolume(poolName string, volName string, h
+	defer revert.Fail()
+
+	// Find the array's qualified name for the configured mode.
+	target, err := d.client().getTarget()
+	if err != nil {
+		return nil, err
+	}
+
+	switch d.config["pure.mode"] {
+	case pureModeISCSI:
+		targetIQN := *target.IQN
+		targetAddr := d.config["pure.array.address"]
+
+		// Try to find an existing iSCSI session.
+		sessionID, err := iscsiFindSession(targetIQN)
+		if err != nil {
+			return nil, err
+		}
+
+		if sessionID != "" {
+			// Already connected to the PureStorage array via iSCSI.
+			// Rescan the session to ensure new volumes are detected.
+			_, err := shared.RunCommand("iscsiadm", "--mode", "session", "--sid", sessionID, "--rescan")
+			if err != nil {
+				return nil, err
+			}
+
+			cleanup := revert.Clone().Fail
+			revert.Success()
+			return cleanup, nil
+		}
+
+		// Attempt to login into discovered iSCSI targets.
+		_, stderr, err := shared.RunCommandSplit(d.state.ShutdownCtx, nil, nil, "iscsiadm", "--mode", "node", "--targetname", targetIQN, "--portal", targetAddr, "--login")
+		if err != nil {
+			return nil, fmt.Errorf("Failed to connect to PureStorage array %q via iSCSI: %s\n%v", targetAddr, stderr, err)
+		}
+	}
+
+	revert.Add(func() { _ = d.disconnect(*target) })
+
+	cleanup := revert.Clone().Fail
+	revert.Success()
+	return cleanup, nil
+}
+
+// disconnect disconnects this host from the given target with a given qualified name.
+func (d *pure) disconnect(target pureTarget) error {
+	switch d.config["pure.mode"] {
+	case pureModeISCSI:
+		targetIQN := *target.IQN
+
+		sessionID, err := iscsiFindSession(targetIQN)
+		if err != nil {
+			return err
+		}
+
+		if sessionID != "" {
+			// Disconnect from the iSCSI target.
+			_, err := shared.RunCommand("iscsiadm", "--mode", "node", "--targetname", targetIQN, "--logout")
+			if err != nil {
+				return fmt.Errorf("Failed disconnecting from PureStorage iSCSI target %q: %w", targetIQN, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// mapVolume maps the given volume onto this host.
+func (d *pure) mapVolume(vol Volume) error {
+	revert := revert.New()
+	defer revert.Fail()
+
+	volName, err := d.getVolumeName(vol)
+	if err != nil {
+		return err
+	}
+
+	unlock, err := locking.Lock(d.state.ShutdownCtx, d.config["pure.mode"])
+	if err != nil {
+		return err
+	}
+
+	defer unlock()
+
+	// Ensure the host exists and is configured with the correct QN.
+	hostname, cleanup, err := d.ensureHost()
+	if err != nil {
+		return err
+	}
+
+	revert.Add(cleanup)
+
+	// Ensure the volume is connected to the host.
+	connCreated, err := d.client().connectHostToVolume(vol.pool, volName, hostname)
+	if err != nil {
+		return err
+	}
+
+	if connCreated {
+		revert.Add(func() { _ = d.client().disconnectHostFromVolume(vol.pool, volName, hostname) })
+	}
+
+	// Connect to the array using iscsi. Connection can be established only
+	// when at least one volume is connected to the host.
+	cleanup, err = d.connect()
+	if err != nil {
+		return err
+	}
+
+	revert.Add(cleanup)
```

Comment:
this can never be called, as you mark the reverter as succeeded on next line?

Reply:
Yes, I wanted to call it anyway to ensure we do not miss the reverter entry if we ever extend the function. Can remove if desired.